{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "32de6955",
   "metadata": {},
   "source": [
    "# Hands-On Session III: Workflows in the workflow collection SPECIMEN\n",
    "In this notebook you will get first-hand experience on how to run some of the steps from the high-quality template-based \n",
    "(HQTB) pipeline within ``SPECIMEN`` along with information on when to use which workflow, how to obtain the required \n",
    "data and how to run the workflows.\n",
    "\n",
    "## Background on workflows\n",
    "Both workflows are intended to be run to obtain a high-quality model with minimal manual input. Hence, both \n",
    "workflows share a similar structure, while the implementation is different.\n",
    "\n",
    "Basic structure\n",
    "- Draft generation\n",
    "- Model extension \n",
    "- Annotation and Clean-up \n",
    "- Analyses (final and after various steps)\n",
    "\n",
    "One major difference between the HQTB and the CarveMe-ModelPolisher-based (CMPB) workflow is that steps in HQTB can be run \n",
    "individually and a template model is required for the draft reconstruction while the CMPB workflow can not only generate \n",
    "a new model but it can also be used to improve an already existing model. \n",
    "\n",
    "### When to use which workflow?\n",
    "Since the structure of the workflow is similar, one could argue, why we need more than one. \n",
    "The table below illustrates this question:\n",
    "\n",
    "|  | **HQTB** | **CMPB** |\n",
    "| :--- | :--- | :--- |\n",
    "| Requirements | Template model, </br> annotated genome from template model organism, </br> annotated genome from organism to be build | Model/ </br> Annotated genome |\n",
    "|Ideally for| Lab strains, </br> where models of close related organisms exist | Novel organisms, </br> without template models/ </br> Existing models |\n",
    "| \"Philosophy\" | Maximise profiting from previous work </br> to prevent wasting time and resources </br> on the same questions again | Automation of \"classic\" workflow |\n",
    "\n",
    "We will take a closer look at the HQTB workflow for this workshop.\n",
    "\n",
    "### Running HQTB\n",
    "Before running HQTB, the necessary input data needs to be gathered, the correct paths added and the options set in \n",
    "one of the possible configuration files for HQTB.\n",
    "\n",
    "#### Getting the input data\n",
    "\n",
    "**Template**\n",
    "\n",
    "Since the HQTB workflow is a template-based approach, one of the first steps is to actually find a suitable template model. \n",
    "For the reconstruction of the model of *Pseudomonas putida* KT2440, the *Pseudomonas aeroginosa* P14 model from Dahal \n",
    "et al.[^1] was chosen (see XML file in `../data/template/`). \n",
    "\n",
    "Due to the bidirectional BLAST in the first part of the workflow, the model itself is not enough for running the workflow.\n",
    "We need the genome file as well. Usually, one would check the model and the corresponding paper for the sequence or identifier for a database where the sequence can be found. \n",
    "Here, we already provide you with the annotated genome used for creating the model, which can be found on NCBI under the assembly number [ASM1462v1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000014625.1/) (see FASTA file in `../data/template/`). \n",
    "\n",
    "**Subject**\n",
    "\n",
    "For the subject organism, the annotated genome is needed as well. It can either be provided directly by the laboratory or downloaded from databases like NCBI. In our case, we will use the NCBI entry [ASM7565v2](https://www.ncbi.nlm.nih.gov/datasets/genome/GCF_000007565.2/) (see FASTA file in `../data/`). For some steps in the workflow, the GFF file is additionally required (either the GenBank, starting with `GCA`, or the RefSeq one, starting with `GCF`).\n",
    "\n",
    "**Other**\n",
    "\n",
    "Depending on how many steps and on what options are enabled, more data files or values are needed, for example:\n",
    "\n",
    "- To set the reaction direction based on BioCyc a table from BioCyc needs to be provided.\n",
    "- To run the KEGGapFiller, a KEGG organism ID is needed.\n",
    "- For running the BioCycGapFiller, one table for the genes, one for the reactions is needed. \n",
    "- The GeneGapFiller requires a DIAMOND database and optionally a mapping table for improving runtime.\n",
    "- etc.\n",
    "\n",
    "### Running steps 1 & 2 for HQTB\n",
    "\n",
    "While the wokflow can be run in a couple of hours or less depending on what options are enabled, it would be too long for the context of this workshop. \n",
    "Therefore, only the first two steps of the workflow will be tested. However, you are encouraged to also generate a full configuration file to get a feeling on what is needed to build a complete model. \n",
    "\n",
    "üìù **Task**\n",
    "1. Fill out the provided example configuration file (see `../data/hqtb_example_config.yaml`).\n",
    "2. Run step 1: Bidirectional BLAST.\n",
    "    > Note: While the workflow can run with both the FASTA and the GBFF, its reliant on matching the feature qualifiers of the GBFF or the ID in the FASTA. Make sure to use the correct one as input.\n",
    "3. Run step 2: Generate draft model.\n",
    "    > Tip: Since the template organism is not of the same species, try out different PID values and see, how they influence the results and why closely related organisms work best for this kind of modelling approach.\n",
    "4. Perform basic analyses on the generated model.\n",
    "\n",
    "<font color=\"grey\">\n",
    "\n",
    "[^1]: Dahal S, Renz A, Dr√§ger A, Yang L. Genome-scale model of Pseudomonas aeruginosa metabolism unveils virulence and drug potentiation. Commun Biol. 2023 Feb 10;6(1):165. doi: 10.1038/s42003-023-04540-8. PMID: 36765199; PMCID: PMC9918512."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f0cdc85",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Your code goes here"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a6f902",
   "metadata": {},
   "source": [
    "<details>\n",
    "<summary>üîë Click to see the answer üîë</summary>\n",
    "\n",
    "Here is the code for the task:\n",
    "\n",
    "```python\n",
    "from specimen.hqtb.core import bidirectional_blast\n",
    "from specimen.hqtb.core import generate_draft_model\n",
    "\n",
    "bidirectional_blast.run(\n",
    "        template = \"./data/template/GCA_000014625.1_ASM1462v1_translated_cds.faa\",\n",
    "        input = \"./data/GCA_000007565.2_ASM756v2_translated_cds.faa\",\n",
    "        dir = \"./hqtb_out/\",  # can be path of your choice\n",
    "    )\n",
    "\n",
    "generate_draft_model.run(\n",
    "    template = \"./data/template/iSD1509.xml\",\n",
    "    bpbbh = \"./hqtb_out/GCA_000007565.2_ASM756v2_translated_cds_GCA_000014625.1_ASM1462v1_translated_cds_bbh.tsv\", # output of the previous function\n",
    "    dir = \"./hqtb_out/\", # can be path of your choice\n",
    "    pid = 90.0,\n",
    "    name = \"iPpuKT2440WS25\")\n",
    "\n",
    "```\n",
    "\n",
    "</details>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa547dd4",
   "metadata": {},
   "source": [
    "## ‚≠êÔ∏è Bonus\n",
    "If you already finished the tasks and still want to go further, you can try to use more options from the configuration \n",
    "file. Otherwise, you can also use more analyses.\n",
    "\n",
    "üìù **Task**\n",
    "\n",
    "Try out more options/steps for HQTB or more analyses.\n",
    "\n",
    "*Note:* More options/steps for HQTB will most likely take more time than the workshop provides."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sprg",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
